Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs

```diff
@@ -41,38 +41,99 @@ public AsyncFunctionAssertions(Func<TTask> subject, IExtractExceptions extractor
-            using var timeoutCancellationTokenSource = new CancellationTokenSource();
-            Task completedTask =
-                await Task.WhenAny(task, Clock.DelayAsync(remainingTime, timeoutCancellationTokenSource.Token));
+            // task already in timeout with sync portion
+            return new AndConstraint<TAssertions>((TAssertions)this);
+        }
 
-            if (completedTask == task)
-            {
-                timeoutCancellationTokenSource.Cancel();
-                await completedTask;
-            }
+        using var timeoutCancellationTokenSource = new CancellationTokenSource();
+        Task completedTask =
+            await Task.WhenAny(task, Clock.DelayAsync(remainingTime, timeoutCancellationTokenSource.Token));
 
-            Execute.Assertion
-                .ForCondition(completedTask == task)
-                .BecauseOf(because, becauseArgs)
-                .FailWith("Expected {context:task} to complete within {0}{reason}.", timeSpan);
+        if (completedTask == task)
+        {
+            timeoutCancellationTokenSource.Cancel();
+            await completedTask;
+        }
+
+        Execute.Assertion
+            .ForCondition(completedTask == task)
+            .BecauseOf(because, becauseArgs)
+            .FailWith("Expected {context:task} to complete within {0}{reason}.", timeSpan);
+
+        return new AndConstraint<TAssertions>((TAssertions)this);
+    }
+
+    /// <summary>
+    /// Asserts that the current <typeparamref name="TTask"/> will complete within the specified time.
+    /// </summary>
+    /// <param name="timeSpan">The allowed time span for the operation.</param>
+    /// <param name="because">
+    /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+    /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+    /// </param>
+    /// <param name="becauseArgs">
+    /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+    /// </param>
+    public async Task<AndConstraint<TAssertions>> NotCompleteWithinAsync(
+        TimeSpan timeSpan, string because = "", params object[] becauseArgs)
+    {
+        bool success = Execute.Assertion
+            .ForCondition(Subject is not null)
+            .BecauseOf(because, becauseArgs)
+            .FailWith("Expected {context:task} to not complete within {0}{reason}, but found <null>.", timeSpan);
+
+        if (!success)
+        {
+            // subject is null, nothing to execute
+            return new AndConstraint<TAssertions>((TAssertions)this);
         }
 
+        ITimer timer = Clock.StartTimer();
+        TTask task = Subject.Invoke();
+        TimeSpan remainingTime = timeSpan - timer.Elapsed;
+        if (remainingTime < TimeSpan.Zero)
+        {
+            // expected timeout reached, done
+            return new AndConstraint<TAssertions>((TAssertions)this);
```

Comment:
🤔 I personally don't like all these intermediate `return` statements, but I don't hold this opinion strongly.

Reply:
I would like to know why do you prefer higher indention instead?
For me, the complexity is lower with early return.
https://docs.codeclimate.com/docs/cognitive-complexity

Reply:
No, I actually prefer to keep the methods small by breaking them down into smaller parts with clear names (without making them too small):

```csharp
    public async Task<AndConstraint<TAssertions>> NotCompleteWithinAsync(
        TimeSpan timeSpan, string because = "", params object[] becauseArgs)
    {
        bool success = Execute.Assertion
            .ForCondition(Subject is not null)
            .BecauseOf(because, becauseArgs)
            .FailWith("Did not expect {context:task} to complete within {0}{reason}, but found <null>.", timeSpan);

        if (success)
        {
            (Task task, TimeSpan executionTime) = InvokeWithTimer(Subject, Clock);
            TimeSpan remainingTime = timeSpan - executionTime;
            if (remainingTime >= TimeSpan.Zero)
            {
                Execute.Assertion
                    .ForCondition(!await CompletesWithinTimeoutAsync(task, remainingTime, Clock))
                    .BecauseOf(because, becauseArgs)
                    .FailWith("Did not expect {context:task} to complete within {0}{reason}.", timeSpan);
            }
        }

        return new AndConstraint<TAssertions>((TAssertions)this);
    }
    
    private static (T result, TimeSpan executionTime) InvokeWithTimer<T>(Func<T> func, IClock clock)
    {
        ITimer timer = clock.StartTimer();
        T result = func.Invoke();
        
        return (result, timer.Elapsed);
    }
    
    private static async Task<bool> CompletesWithinTimeoutAsync(Task target, TimeSpan remainingTime, IClock clock)
    {
        using var timeoutCancellationTokenSource = new CancellationTokenSource();
              
        Task completedTask =
            await Task.WhenAny(target, clock.DelayAsync(remainingTime, timeoutCancellationTokenSource.Token));

        if (completedTask == target)
        {
            timeoutCancellationTokenSource.Cancel();
            await completedTask;
            return true;
        }
        
        return false;
    }
```

And I would make `InvokeWithTimer` and `CompletesWithinTimeoutAsync` extension methods.

Reply:
Personally I also praise return-early-return-often, but for Fluent Assertions I try to follow Dennis' preference of single returns in the name of consistency.

The few exceptions (that I can easily find) are places where using single return would require additional variables/work.
E.g. when returning early also act as breaking early out of a loop.

https://github.com/fluentassertions/fluentassertions/blob/de9faf3057b261ce4658a4579d8a62a2e811e2f3/Src/FluentAssertions/Collections/GenericCollectionAssertions.cs#L2248-L2288

In any case I would really prefer to keep style changes separated from functional changes 🙏

Reply:
@jnyrup did you see my suggested refactoring bove?

Reply:
I did and it looks fine to me. 
It attacks some underlying problems from other angles.

1. Having too many active variables at a time.
2. A method being so lengthy that it requires long vertical eye travel or even scrolling to get an overview of.

Reply:
Based on your suggestion, Dennis, I reworked the methods using new helper methods.
I did not create them as extension methods because:
What kind if type should be extended? The relevant types are very generic. Extension methods (should be public for extensibility) must not accidentally used in customers code.

I understand the idea, Dennis, to keep methods small. But this is not an argument for single-return.
So far I will go with Jonas and accept the environment style.

Reply:
Great to see you liked my suggestions. And for the record, returning out of the method at the beginning, based on some precondition, is totally fine by me as well. @jnyrup and I are completely aligned on this.